### DISCUSSION TITLE
Digest Authentication

### DISCUSSION THREAD

**Discussion by sundarv85** [go-resty/resty#467] [Q&A]
I am trying to use Digest Authentication for the URL: `https://httpbin.org/digest-auth/auth/user/password`

```	
resp, err := client.R().
		EnableTrace().
                SetBasicAuth("user", "password").
		SetAuthScheme("Digest").
		Get("https://httpbin.org/digest-auth/auth/user/password")
```

I suppose that is not the correct way to pass the username/password. Any suggestions?

**Comment by jeevatkm** [comment]
@sundarv85 what do you mean? do you see an alternate way, if yes then please share it here?
> I suppose that is not the correct way to pass the username/password

Typically, you should not hardcode username and password in the version control. It is recommended to use environment variables or settings files to store credentials (e.g.: if AWS is used then you can rely on Secret manager, SSM), and then grab those values dynamically and supply it to the method.

**Reply by sundarv85** [reply]
Hi @jeevatkm 

My question probably was not clear. I'm trying to get the digest authentication to work with go-resty (basic auth works with Resty). It is not about how to store the username/password. That was just the example I had given.
 
For e.g. httpbin for digest authentication example URL is:

https://httpbin.org/digest-auth/auth/user/password

If you open the above link, you will see that it works with user/password combination. However, when I use it as I have described below, then it does not work with Resty.

```
package main

import (
	"fmt"
	"github.com/go-resty/resty/v2"
)

func main() {
	// Create a Resty Client
	client := resty.New()

	resp, err := client.R().
		EnableTrace().
		SetBasicAuth("user", "password").
		SetAuthScheme("Digest").
		Get("https://httpbin.org/digest-auth/auth/user/password")
}
```

I get the output

```
Response Info:
  Error      : <nil>
  Status Code: 401
  Status     : 401 Unauthorized
  Proto      : HTTP/2.0
  Time       : 547.399421ms
  Received At: 2021-10-26 09:40:38.529841152 +0200 CEST m=+0.547902712
  Body       :
```

So I'm not sure how to get digest authentication to work with resty.. with resty I'm able to get both basic and bearer to work, just the digest is not working..

**Comment by jeevatkm** [comment]
@sundarv85 Now I understand your question. Thanks. 
As of now, the Resty library does not provide OOTB helper functions for `digest` authentication. However, you can use any helper library for creating Digest auth header and use it with Resty.

Maybe I will see if I can add supported functions in the future.

**Reply by sundarv85** [reply]
Thanks @jeevatkm  - Do you have any specific library recommendation that I can use to create the digest auth header..

Of course, it would be awesome to have that as a feature within go-resty! :+1:

**Reply by moorereason** [reply]
I'd suggest https://github.com/icholy/digest

**Comment by segevda** [comment]
I have created a [PR ](https://github.com/go-resty/resty/pull/583) to add OOTB digest authentication support, I would highly appreciate if you can review :)

cc: @jeevatkm
